### DISCUSSION TITLE
Global setup/teardown functions for all tests in watch mode

### DISCUSSION THREAD

**Discussion by colinlienard** [vitest-dev/vitest#6719] [Q&A] [answered]
In Vitest, we have several options for running setup and teardown functions:

1. `beforeAll` and `afterAll` in setup files which run once for each test file
2. Global setup files with `setup` and `teardown` functions which run once when the process starts and stops

I'm looking for a way to run specific functions at the very beginning and end of all tests, like with a global setup, but also in watch mode, meaning:
   - When a file changes or I press "r", the setup function should run again
   - Tests run
   - Finally, the teardown function should run

Is there an existing feature in Vitest that can achieve this behavior?
If not, would this be considered a valuable feature to add to Vitest in the future?

(I found this [discussion](https://github.com/vitest-dev/vitest/discussions/6233) that seems to ask the same question as me, but got no answers)

**Comment by sheremet-va** [comment, marked as answer]
Currently the only way to run a custom code when Vitest triggers a rerun is with a [custom reporter's](https://github.com/vitest-dev/vitest/issues/6417#issuecomment-2337230504) `onWatcherRerun`. We are planning to add the same hook to the global setup in the future: https://github.com/vitest-dev/vitest/issues/6417#issuecomment-2337230504

**Reply by colinlienard** [reply]
Thanks, that does the trick

For anyone passing by, here is what my reporter looks like:

```ts
export default class CustomReporter implements Reporter {
	async onInit() {
		await setup();
	}
	async onWatcherRerun() {
		await setup();
	}
	async onFinished() {
		await teardown();
	}
}
```
